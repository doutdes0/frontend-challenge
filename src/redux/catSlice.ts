import { createSlice } from "@reduxjs/toolkit";
import { PayloadAction } from "@reduxjs/toolkit";
//import { getAllUsers } from "../thunks";

interface CatsState {
  [id: string]: {
    id: string;
    url: string;
    isLiked: boolean;
  };
}

const initialState: CatsState = {};

const catsSlice = createSlice({
  name: "cats",
  initialState,
  reducers: {
    toggle_like: (state, action: PayloadAction<string>) => {
      state[action.payload].isLiked = !state[action.payload].isLiked;
    },
  },
});

export default catsSlice.reducer;
