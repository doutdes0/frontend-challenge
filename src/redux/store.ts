import { configureStore } from "@reduxjs/toolkit";
import catsReducer from "./catSlice";

export const store = configureStore({
  reducer: {
    cats: catsReducer,
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      serializableCheck: false,
    }),
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
