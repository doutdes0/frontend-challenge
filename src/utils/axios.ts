import axios from "axios";

export const catnipThrower = axios.create({
  baseURL: " https://api.thecatapi.com/v1/images/search",
  params: {
    limit: 15,
    api_key: `${process.env.VITE_API_KEY}`,
  },
});
