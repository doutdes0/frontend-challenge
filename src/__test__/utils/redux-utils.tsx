import { setUpStore } from "../../redux/store";
import { Provider } from "react-redux";
import { AppStore, RootState } from "../../redux/store";
import { render, RenderOptions } from "@testing-library/react";

interface ExtendedRenderOptions extends Omit<RenderOptions, "queries"> {
  preloadedState?: Partial<RootState>;
  store?: AppStore;
}

export const renderWithProvider = (
  component: React.ReactElement,
  {
    preloadedState = {},
    store = setUpStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) => {
  const _provider = ({ children }: { children: React.ReactNode }) => {
    return <Provider store={store}>{children}</Provider>;
  };
  return {
    store,
    ...render(component, { wrapper: _provider, ...renderOptions }),
  };
};

export const key = "67b3f85d-5354-4bd6-ba9a-5482fb966057";
export const preloadedState: RootState = {
  cats: {
    "67b3f85d-5354-4bd6-ba9a-5482fb966057": {
      id: "67b3f85d-5354-4bd6-ba9a-5482fb966057",
      url: "https://cdn2.thecatapi.com/images/2n3.jpg",
      isLiked: false,
    },
  },
};
