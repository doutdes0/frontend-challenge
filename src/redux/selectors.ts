import { createSelector } from "@reduxjs/toolkit";
import { RootState } from "./store";

const selectCats = (state: RootState) => state.cats;

export const selectCatsValues = createSelector(selectCats, (cats) =>
  Object.values(cats)
);
export const selectFavoriteCats = createSelector(selectCats, (cats) => {
  const values = Object.values(cats);
  const favorites = values.filter((cat) => cat.isLiked);
  return favorites;
});
