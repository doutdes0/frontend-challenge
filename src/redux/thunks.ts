import { catnipThrower } from "../utils/axios";
import { createAsyncThunk } from "@reduxjs/toolkit";
import axios from "axios";

type APIRes = {
  data: {
    id: string;
    url: string;
    width: number;
    height: number;
    breeds?: [];
  }[];
};

export const getCats = createAsyncThunk("meow", async (_, thunkAPI) => {
  try {
    const res = await catnipThrower.get<any, APIRes>("");
    const data = res.data.map((cat) => cat.url);
    return data;
  } catch (e) {
    if (axios.isAxiosError(e)) {
      return thunkAPI.rejectWithValue(e.response?.data.msg);
    } else {
      return thunkAPI.rejectWithValue(e);
    }
  }
});
