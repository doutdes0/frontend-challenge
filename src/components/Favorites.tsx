import styled from "styled-components";
import { useAppSelector } from "../hooks/useRedux";
import Card from "./Card";
import { selectFavoriteCats } from "../redux/selectors";

const Favorites: React.FC = () => {
  const favorites = useAppSelector(selectFavoriteCats);

  return (
    <Container>
      {favorites.length > 0 &&
        favorites.map(({ id, url, isLiked }) => (
          <Card
            key={id}
            id={id}
            isLiked={isLiked}
            url={url}
          />
        ))}
    </Container>
  );
};

export default Favorites;

const Container = styled.div`
  position: relative;
  padding: 62px 62px 124px;
  display: grid;
  grid-template-columns: repeat(auto-fill, 225px);
  grid-gap: 2rem;
  justify-content: space-between;
`;
