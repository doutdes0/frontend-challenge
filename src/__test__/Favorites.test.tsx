import Favorites from "../components/Favorites";
import { BrowserRouter } from "react-router-dom";
import { renderWithProvider, preloadedState } from "./utils/redux-utils";
import { screen } from "@testing-library/react";
import { toggle_like } from "../redux/catSlice";
import { act } from "react-dom/test-utils";

describe("Favotites", () => {
  it("should match snapshot", () => {
    const { container } = renderWithProvider(
      <BrowserRouter>
        <Favorites />
      </BrowserRouter>,
      { preloadedState }
    );
    expect(container).toMatchSnapshot();
  });
  it("should not render unfavorited cards", () => {
    renderWithProvider(
      <BrowserRouter>
        <Favorites />
      </BrowserRouter>,
      { preloadedState }
    );
    expect(screen.queryByAltText("cat")).not.toBeTruthy();
  });
  it("should render favorited cards", () => {
    const { store } = renderWithProvider(
      <BrowserRouter>
        <Favorites />
      </BrowserRouter>,
      { preloadedState }
    );
    act(() => {
      store.dispatch(toggle_like(Object.keys(preloadedState.cats)[0]));
    });
    expect(screen.getByAltText("cat")).toBeTruthy();
  });
});
