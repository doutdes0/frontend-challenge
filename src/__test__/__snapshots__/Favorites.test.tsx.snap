// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Favotites should match snapshot 1`] = `
<div>
  <div
    class="sc-dmyCSP knOGob"
  />
</div>
`;
